sdk/core/azure-core/azure/core/credentials.py

```diff
@@ -59,36 +59,36 @@ class AzureKeyCredential:
     """Credential type used for authenticating to an Azure service.
     It provides the ability to update the key without creating a new client.
 
-    :param str key: The key used to authenticate to an Azure service
+    :param Union[str, Tuple[str, ...]] key: The key used to authenticate to an Azure service
     :raises: TypeError
     """
 
-    def __init__(self, key: str) -> None:
-        if not isinstance(key, str):
-            raise TypeError("key must be a string.")
+    def __init__(self, key: Union[str, Tuple[str, ...]]) -> None:
```

Comment:
I really don't like the tuple story personally, as it means nothing is typed and people needs to know in what order put the information. If we mess this up and people reverse the headers, the call becomes wrong. Now as well `key` (singular) may return two keys, which makes no sense.

Sorry don't like it at all, but I defer to @annatisch adn @johanste

Reply:
We should make the class generic. This should (at least partly) address @lmazuel's concerns.

Reply:
...and to expand on my comment; you can constrain the key parameter to a named tuple - or event a typed dict - if you want to associate names with each key. Or you can provide a concrete class implementing the protocol and only allow that in your constructor.

Ultimately, the key credential policy and the key credential are going to have to agree on the protocol between them. Today that is implicit (but simple because there is but one key allowed and the policy carries the one and only header). A case can be made that the header name should always have been carried by the credential (because what are the odds that the same key can be used for two different services using two different headers).

What I'd like to avoid is to have a dummy key credential class that has `header1`/`header2`/.../`headerN` properties.

Reply:
How about this one? 

https://github.com/Azure/azure-sdk-for-python/pull/28754

@johanste @annatisch @lmazuel